lib/duniverse.ml

```diff
@@ -1,15 +1,148 @@
+      in
+      let new_ref =
+        match Ordering.of_int (OpamVersionCompare.compare t.ref package.ref) with
+        | Gt | Eq -> t.ref
+        | Lt -> package.ref
+      in
+      { t with
+        dir = new_dir;
+        ref = new_ref;
+        provided_packages = package.opam :: t.provided_packages
+      }
+
+    let aggregate_packages l =
+      let update map ({ Package.upstream; _ } as package) =
+        String.Map.update map upstream ~f:(function
+          | None -> Some (from_package package)
+          | Some t -> Some (aggregate t package) )
+      in
+      let aggregated_map = List.fold_left ~init:String.Map.empty ~f:update l in
+      String.Map.values aggregated_map
+  end
+
+  module One = struct
+    type t = Opam of Opam.t | Source of Source.Package.t
+
+    let equal t t' =
+      match (t, t') with
+      | Opam opam, Opam opam' -> Opam.equal opam opam'
+      | Source source, Source source' -> Source.Package.equal source source'
+      | (Opam _ | Source _), _ -> false
+
+    let raw_pp fmt t =
+      match t with
+      | Opam opam -> Format.fprintf fmt "@[<hov 2>Opam@ %a@]" Opam.raw_pp opam
+      | Source source -> Format.fprintf fmt "@[<hov 2>Source@ %a@]" Source.Package.raw_pp source
+
+    let from_opam_entry ~get_default_branch entry =
+      let open Types.Opam in
+      let open Result.O in
+      match entry with
+      | { dev_repo = `Virtual; _ } | { dev_repo = `Error _; _ } -> Ok None
+      | { is_dune = false; package = { name; version }; _ } -> Ok (Some (Opam { name; version }))
+      | { is_dune = true; dev_repo = `Git upstream; tag = Some ref; package = { name; version } }
+        ->
+          Ok (Some (Source { opam = { name; version }; upstream; ref }))
+      | { is_dune = true; dev_repo = `Git upstream; tag = None; package = { name; version } } ->
+          get_default_branch upstream >>= fun ref ->
+          Ok (Some (Source { opam = { name; version }; upstream; ref }))
+
+    let partition_list l =
+      List.partition_map ~f:(function Opam o -> Left o | Source s -> Right s) l
+  end
+
+  type t = { opamverse : Opam.t list; duniverse : Source.t list } [@@deriving sexp]
+
+  let equal t t' =
+    List.equal Opam.equal t.opamverse t'.opamverse
+    && List.equal Source.equal t.duniverse t'.duniverse
+
+  let raw_pp fmt t =
+    let open Pp_combinators.Ocaml in
+    Format.fprintf fmt "@[<hov 2>{ opamverse = %a;@ duniverse = %a}@]" (list Opam.raw_pp)
+      t.opamverse (list Source.raw_pp) t.duniverse
+
+  let from_opam_entries ~get_default_branch entries =
+    let open Result.O in
+    let results = List.map ~f:(One.from_opam_entry ~get_default_branch) entries in
+    Result.List.all results >>= fun dep_options ->
+    let deps = List.filter_opt dep_options in
+    let opamverse, source_deps = One.partition_list deps in
+    let duniverse = Source.aggregate_packages source_deps in
+    Ok { opamverse; duniverse }
```

Comment:
What about moving this 3 last lines to a `from_ones` (or `from_classified`) and nesting `One.partition_list` and `Source.aggregate_packages` ?

To split better the classify logic and the pure construction of the structure. Also, these functions do not really fit in their modules.

Reply:
Hmm yeah I think I see what you mean, I'll give it a go!